torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1820,6 +1831,896 @@ def clip_grad_norm_(
                 assert p.grad is not None
                 p.grad.detach().mul_(clip_coef.to(p.grad.device))
 
+    @staticmethod
+    def full_optim_state_dict(
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        optim_input: Optional[Union[
+            List[Dict[str, Any]], Iterable[torch.nn.Parameter],
+        ]] = None,
+        target_rank: int = 0,
+    ) -> Dict[str, Any]:
+        """
+        Consolidates the full optimizer state on rank ``target_rank`` and
+        returns it as a :class:`dict` following the convention of normal
+        PyTorch optimizers. This involves remapping from flattened to
+        unflattened parameters for any FSDP modules contained in ``model`` so
+        that the state is agnostic to the current world size and resulting
+        sharding scheme.
+
+        The returned :class:`dict` is identical to the result of
+        ``torch.optim.Optimizer.state_dict()`` for an optimizer constructed
+        using the parameters of a purely non-wrapped version of the model (i.e.
+        without any ``FullyShardedDataParallel`` modules).
+
+        Args:
+            model (torch.nn.Module): Root module (which may or may not be a
+                :class:`FullyShardedDataParallel` instance) whose parameters
+                were passed into the optimizer ``optim``.
+            optim (torch.optim.Optimizer): Optimizer for ``model`` 's
+                parameters.
+            optim_input (Optional[Union[List[Dict[str, Any]],
+            Iterable[torch.nn.Parameter]]]): Input passed into the optimizer
+                ``optim`` representing either a :class:`list` of parameter
+                groups or an iterable of parameters; if ``None``, then this
+                method assumes the input was ``model.parameters()``. (Default:
+                ``None``)
+            target_rank (int): Rank on which to save the full optimizer state
+                dict in CPU memory. (Default: 0)
```

Comment:
Do we want to offload to CPU by default? 

Eventually, we should consolidate this rank 0 only + CPU vs GPU saving with state_dict, which will support rank 0 only + offload_to_cpu as well (through summon_full_params).